docs/architecture/adr-050-auto-compound-rewards.md

```diff
@@ -0,0 +1,167 @@
+# ADR 050: Auto-Compound Rewards
+
+## Changelog
+
+* April 17, 2022: Initial Draft
+
+## Status
+
+Draft (not implemented)
+
+## Abstract
+
+This ADR describes a modification of the `x/distribution` module's functionality
+to allow users to request the ability to auto-compound their rewards to their
+delegated validators on-chain.
+
+## Context
+
+As of SDK version v0.45.x, the `x/distribution` module defines a mechanism in
+which delegators receive rewards by delegating voting power to validators in the
+form of a native staking token. The reward distribution itself happens in a lazy
+fashion and is defined by the [F1 specification](https://drops.dagstuhl.de/opus/volltexte/2020/11974/pdf/OASIcs-Tokenomics-2019-10.pdf).
+In other words, delegators accumulate "unrealized" rewards having to explicitly
+execute message(s) on-chain in order to withdraw said rewards. This provides the
+ability for the chain to not have to explicitly distribute rewards to delegators
+on a block-by-block basis which would otherwise make the network crawl to halt
+as the number of delegations increases over time.
+
+However, it has been shown that there is a strong desire to auto-compound
+distribution rewards. As a result, there has been a proliferation of scripts, tooling
+and clients off-chain to facilitate such a mechanism. However, these methods are
+ad-hoc, often provide a cumbersome UX, and do not scale well to multiple networks.
+
+Thus, we propose a mechanism to modify the `x/distribution` module to allow for
+delegators to auto-compound their rewards on-chain.
+
+## Decision
+
+In order to facilitate auto-compounding rewards, we need the ability for delegators
+to opt into having their rewards auto-compounded. There are numerous ways to approach
+this, where a simple method would to introduce a new message type, `MsgEnableAutoCompoundRewards`,
+defined as follows:
+
+```protobuf
+message MsgEnableAutoCompoundRewards {
```

Comment:
a huge feature could be to add a percentage here. (defaults to 100%) The problem with the current form of auto compounding is the user cant realise any of their rewards and must stop auto compounding then wait a few days. The percentage here would say only claim and relegate a % of the rewards